core/include/vecmem/memory/binary_page_memory_resource.hpp

```diff
@@ -0,0 +1,159 @@
+/**
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include "vecmem/memory/resources/memory_resource.hpp"
+
+#include <cstddef>
+#include <memory>
+#include <vector>
+
+namespace vecmem {
+    /**
+     * @brief A memory manager using power-of-two pages that can be split to
+     * deal with allocation requests of various sizes.
+     *
+     * This is a non-terminal memory resource which relies on an upstream
+     * allocator to do the actual allocation. The allocator will allocate only
+     * large blocks with sizes power of two from the upstream allocator. These
+     * blocks can then be split in half and allocated, split in half again. This
+     * creates a binary tree of pages which can be either vacant, occupied, or
+     * split.
+     */
+    class binary_page_memory_resource : public memory_resource {
+    public:
+        /**
+         * @brief Initialize a binary page memory manager depending on an
+         * upstream memory resource.
+         */
+        binary_page_memory_resource(memory_resource *);
```

Comment:
I prefer using references in functions that must receive an argument. The member variable should still be a pointer, as that is a lot easier on the compiler for generating copy/move constructors/assignment operators automatically. But here you should receive the upstream `memory_resource` rather as a reference.

Reply:
I totally agree with you. I went for a pointer argument here because that's what the standard library does, but I honestly have no idea why.

Reply:
I think the PMR implementation can accept `nullptr` as the memory resource. In which case it would fall back on some standard memory resource. (Which it manages through those helper functions that you pointed out earlier.)

So in their design the pointer value makes sense. (I think...) But in this project, as we discussed before, let's not bother with default memory resources...